native/packages/react-native-bpk-component-button/src/BpkButton.js

```diff
@@ -16,215 +16,269 @@
-       theme = null;
-     }
-   }
-
-   const accessibilityTraits = ['button'];
-   if (disabled) {
-     accessibilityTraits.push('disabled');
-   }
-
-   // Note that TouchableHighlight isn't on Android, so TouchableFeedback
-   // will need to be used to support it.
-   return (
-     <LinearGradient
-       colors={getGradientColors(theme, props)}
-       style={[getStyleForElement('container', props), style]}
-     >
-       <TouchableHighlight
-         style={[getStyleForElement('button', props), getThemingForElement('button', theme, props)]}
-         disabled={disabled}
-         onPress={onPress}
-         underlayColor={styles.underlayColor}
-         accessibilityComponentType="button"
-         accessibilityLabel={accessibilityLabel || title}
-         accessibilityTraits={accessibilityTraits}
-         {...rest}
-       >
-         <View style={getStyleForElement('view', props)}>
-           { !iconOnly &&
-             <BpkText
-               textStyle={large ? 'lg' : 'sm'}
-               emphasize
-               style={[getStyleForElement('text', props), getThemingForElement('text', theme, props)]}
-             >
-               {title}
-             </BpkText>
-           }
-           {icon}
-         </View>
-       </TouchableHighlight>
-     </LinearGradient>
-   );
- };
-
- const propTypes = {
-   title: PropTypes.string.isRequired,
-   icon: iconPropType,
-   iconOnly: PropTypes.bool,
-   onPress: PropTypes.func.isRequired,
-   type: PropTypes.oneOf(BUTTON_TYPES),
-   large: PropTypes.bool,
-   disabled: PropTypes.bool,
-   accessibilityLabel: PropTypes.string,
-   style: View.propTypes.style,
-   theme: themePropType,
- };
-
- BpkButton.propTypes = propTypes;
-
- BpkButton.defaultProps = {
-   icon: null,
-   iconOnly: false,
-   type: 'primary',
-   large: false,
-   disabled: false,
-   accessibilityLabel: null,
-   style: null,
-   theme: null,
- };
-
- export default withTheme(BpkButton);
- export {
-   propTypes,
-   BUTTON_TYPES,
-  };
+  let theme = props.theme;
+
+  // Validate the button type.
+  if (!BUTTON_TYPES.includes(type)) {
+    throw new Error(`"${type}" is not a valid button type. Valid types are ${BUTTON_TYPES.join(', ')}`);
+  }
+
+  // Validate that button is themeable and all theming attributes
+  // have been supplied. If not, disable theming.
+  if (theme) {
+    if (!isTypeThemeable(type) || !themeAttributesSupplied(type, theme)) {
+      theme = null;
+    }
+  }
+
+  const accessibilityTraits = ['button'];
+  if (disabled) {
+    accessibilityTraits.push('disabled');
+  }
+  /**
+   * function to get the iOS container for the button
+   * it's declared here rather than outside the component
+   * so it has access to the lexical scope of the component
+   * and all the prop, theme etc is available
+   *
+   * @param {ReactElement} content button's content
+   * @returns {ReactElement} container iOS specific container
+   */
+  function getIOSContainer(content) {
```

Comment:
Why are we using `function` over arrow function + const var here?

Reply:
It's just personal choice, keeps it more functional and will leverage hoisting (it's declared on top anyway tough)

Reply:
Okay I guess it's down to persoanl choice. I would have used const + arrow function. Hoisting is a bad thing in my head

Reply:
I disagree on hoisting being a bad thing, but for consistency's sake I've updated to use const + arrow function